### PR TITLE
Change RaftSpare node to RaftVoter

### PR DESCRIPTION
**Pull request by ichiro999**
Sometime，In a 2-node cluster that one node's role is RaftSpace and the ohter one is Leader, I Remove the leader node from the cluster; At this time, Remove Operation was failed and the cluster would be unavailable, because of no node can be Leader.

So I change the RaftSpare node to RaftVote when remove leader from 2-node cluster

### CODE REVIEW

**Comment by tomponline** [comment]
Originally from https://github.com/canonical/lxd/pull/12035

**Comment by nanjj** [comment]
> This change makes sense to me.

I found one usage scenario to reproduce the problem:
1. Create a 6 nodes lxd cluster, say:
    | Node | Role |
    |--------|------|
    | lxd01  | Database-Leader |
    | lxd02  | Database |
    | lxd03 | Database |
    | lxd04 | Standby |
    | lxd05 | Standby |
    | lxd06 |  |
2. Stop lxd06, or shutdown lxd06,
3. lxc cluster remove lxd02 lxd03 lxd04 lxd05,
4. Start lxd06, now the cluster list:
    | Node | Role |
    |--------|------|
    | lxd01 | Database-Leader |
    | lxd06 |  |
5. Now `lxc cluster remove lxd01` will trigger the bug and apply the fix in the PR can fix the problem.

**Review by tomponline — CHANGES_REQUESTED** [review]
Thanks for this!

Please can you update the `test_clustering_remove_members` tests to test for this change.

Thanks

File: test/suites/clustering.sh [hunk 1]
```diff
@@ -3228,66 +3228,122 @@ test_clustering_remove_members() {
...
-
-  daemon_pid2=$(cat "${LXD_TWO_DIR}/lxd.pid")
   shutdown_lxd "${LXD_TWO_DIR}"
-
-  daemon_pid3=$(cat "${LXD_THREE_DIR}/lxd.pid")
   shutdown_lxd "${LXD_THREE_DIR}"
+  shutdown_lxd "${LXD_FOUR_DIR}"
+  shutdown_lxd "${LXD_FIVE_DIR}"
+
+  # Force-kill the last two to prevent leadership loss.
```
Comment (tomponline) [inline test/suites/clustering.sh:3321]: Could you expand the comment as to why this is needed please?

File: test/suites/clustering.sh [hunk 2]
```diff
@@ -3228,66 +3228,122 @@ test_clustering_remove_members() {
...
+  LXD_DIR="${LXD_SIX_DIR}" lxc cluster rm node1
 
-  # Ensure only node2 is left in the cluster
-  LXD_DIR="${LXD_TWO_DIR}" lxc cluster ls | grep -q node2
-  ! LXD_DIR="${LXD_TWO_DIR}" lxc cluster ls -f csv | grep -qv node2 || false
+  # Ensure the remaining node is working
+  ! LXD_DIR="${LXD_SIX_DIR}" lxc cluster list | grep -q "node1" || false
+  LXD_DIR="${LXD_SIX_DIR}" lxc cluster list | grep -q "node6"
 
-  # Ensure clustering is disabled (no output) on node1 and node3
```
Comment (tomponline) [inline test/suites/clustering.sh:3263]: We seem to have lost some scenarios that were being tested before is this due to the change in logic?

File: test/suites/clustering.sh [hunk 3]
```diff
@@ -3228,66 +3228,122 @@ test_clustering_remove_members() {
...
-  spawn_lxd_and_join_cluster "${ns3}" "${bridge}" "${cert}" 3 2 "${LXD_THREE_DIR}"
+  # make sure node 6 is spare
+  LXD_DIR="${LXD_ONE_DIR}" lxc cluster list | grep -q "node6"
+  ! LXD_DIR="${LXD_ONE_DIR}" lxc cluster show node6 | grep -qE "\- database-standy$|\- database-leader$|\- database$" || false
 
-  # Ensure successful communication
-  LXD_DIR="${LXD_TWO_DIR}" lxc info --target node3 | grep -q "server_name: node3"
-  LXD_DIR="${LXD_THREE_DIR}" lxc info --target node2 | grep -q "server_name: node2"
+  LXD_DIR="${LXD_ONE_DIR}" lxc cluster list
+  # waite for new leader ready
```
Comment (tomponline) [inline test/suites/clustering.sh:3290]: Typo

File: test/suites/clustering.sh [hunk 4]
```diff
@@ -3228,66 +3228,122 @@ test_clustering_remove_members() {
   ns2="${prefix}2"
   spawn_lxd_and_join_cluster "${ns2}" "${bridge}" "${cert}" 2 1 "${LXD_TWO_DIR}"
 
-  # Ensure successful communication
+  # Spawn a three node
```
Comment (tomponline) [inline test/suites/clustering.sh:3231]: Please can you put these test changes in a separate commit.

File: test/suites/fdleak.sh [hunk 5]
```diff
@@ -40,7 +40,7 @@ test_fdleak() {
 
   bad=0
   # shellcheck disable=SC2015
-  [ ${leakedfds} -gt 5 ] && bad=1 || true
+  [ "${leakedfds}" -gt 5 ] && bad=1 || true
```
Comment (tomponline) [inline test/suites/fdleak.sh:43]: This should be a separate commit please.

File: test/suites/migration.sh [hunk 6]
```diff
@@ -72,6 +72,7 @@ test_migration() {
 
   lxc_remote remote remove l1
   lxc_remote remote remove l2
+  shutdown_lxd "${LXD2_DIR}"
```
Comment (tomponline) [inline test/suites/migration.sh:75]: Whybis this change needed?

Also this should be a separate commit please.

**Review by tomponline — CHANGES_REQUESTED** [review]
Thanks for this, I've made some comments.

File: test/suites/clustering.sh (hunk 2 above)
Comment (ichiro999) [reply test/suites/clustering.sh:3263]: This test is to construct a scenario where a cluster only have a spare node and leader node，then remove the leader，check other node become to leader and cluster is still work by join a new node，so i think my test include the scenarios what old test caase。

File: test/suites/fdleak.sh (hunk 5 above)
Comment (ichiro999) [reply test/suites/fdleak.sh]: done

File: test/suites/clustering.sh (hunk 3 above)
Comment (ichiro999) [reply test/suites/clustering.sh:3289]: done

File: test/suites/clustering.sh [hunk 7]
```diff
@@ -3228,66 +3228,122 @@ test_clustering_remove_members() {
...
-  LXD_DIR="${LXD_TWO_DIR}" lxc cluster ls | grep -q node2
-  ! LXD_DIR="${LXD_TWO_DIR}" lxc cluster ls -f csv | grep -qv node2 || false
+  # Ensure the remaining node is working
+  ! LXD_DIR="${LXD_SIX_DIR}" lxc cluster list | grep -q "node1" || false
+  LXD_DIR="${LXD_SIX_DIR}" lxc cluster list | grep -q "node6"
 
-  # Ensure clustering is disabled (no output) on node1 and node3
-  ! LXD_DIR="${LXD_ONE_DIR}" lxc cluster ls | grep -qv "." || false
-  ! LXD_DIR="${LXD_THREE_DIR}" lxc cluster ls | grep -qv "." || false
+  #make sure node6 is leader and database
```
[the remark this answers is not in the document]
Comment (ichiro999) [reply test/suites/clustering.sh]: done

**Review by tomponline — APPROVED** [review]
Thanks!
